fuel-vm/src/interpreter/flow.rs

```diff
@@ -147,138 +159,392 @@ where
     Tx: ExecutableTransaction,
 {
     fn _prepare_call(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<(), RuntimeError> {
-        let (ax, overflow) = a.overflowing_add(32);
-        let (cx, of) = c.overflowing_add(32);
-        let overflow = overflow || of;
-
-        if overflow || ax > VM_MAX_RAM || cx > VM_MAX_RAM {
-            return Err(PanicReason::MemoryOverflow.into());
+        let params = PrepareCallParams {
+            call_params_mem_address: a,
+            amount_of_coins_to_forward: b,
+            asset_id_mem_address: c,
+            amount_of_gas_to_forward: d,
+        };
+        let current_contract = if self.context.is_internal() {
+            Some(*internal_contract(
+                &self.context,
+                self.registers.fp(),
+                self.memory.as_ref(),
+            )?)
+        } else {
+            None
+        };
+        let memory = PrepareCallMemory::try_from((self.memory.as_mut(), &params))?;
+        let input_contracts = self.tx.input_contracts().copied().collect();
+
+        PrepareCallInput {
+            params,
+            registers: (&mut self.registers).into(),
+            memory,
+            context: &mut self.context,
+            gas_cost: self.gas_costs.call,
+            runtime_balances: &mut self.balances,
+            storage: &mut self.storage,
+            input_contracts,
+            panic_context: &mut self.panic_context,
+            receipts: &mut self.receipts,
+            script: self.tx.as_script_mut(),
+            consensus: &self.params,
+            frames: &mut self.frames,
+            current_contract,
+            profiler: &mut self.profiler,
         }
+        .prepare_call()
+    }
 
-        let call = Call::try_from(&self.memory[a as usize..])?;
-        let asset_id =
-            AssetId::try_from(&self.memory[c as usize..cx as usize]).expect("Unreachable! Checked memory range");
+    /// Prepare a call instruction for execution
+    pub fn prepare_call(&mut self, ra: RegId, rb: RegId, rc: RegId, rd: RegId) -> Result<(), RuntimeError> {
+        const M: &str = "the provided id is not a valid register";
 
-        let mut frame = self.call_frame(call, asset_id)?;
+        let a = self
```

Comment:
Maybe we can put it into inner closure or a method not to duplicate the same code?

Reply:
This isn't from this PR